---
title: "raw_data_analysis"
output:
  html_document: default
  pdf_document: default
date: "2022-09-21"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = '/Users/cgu3/Library/CloudStorage/OneDrive-InsideMDAnderson/data/missing-protein-project')
setwd('/Users/cgu3/Library/CloudStorage/OneDrive-InsideMDAnderson/data/missing-protein-project')
getwd()
```


```{r}
library(tidyverse)
# install.packages("devtools")
if (!require("protools", quietly = TRUE))
   devtools::install_github("https://github.com/FDUguchunhui/protools")
library('protools')
```


# Import data
```{r}
# import gastric primary cell nsaf data that filtered with Spectral count (SpC) >=2
gastric_nsaf <- as.matrix(read.csv('data/gastric-primary-cell/gastric_primary_cell_nsaf_2.csv', row.names = 1))
# # import gastric primary cell SpC data
gastric_count <- read_csv('data/gastric-primary-cell/gastric_primary_cell_count_2.csv', show_col_types = FALSE)
# rename the index column as "accession"
colnames(gastric_count)[1] <- 'accession'
# import information for mapping between accession, gene symbol, and ensembl ID
omics_metadata <- read_csv('data/support-data/omics_metadata.csv', show_col_types = FALSE)
# load missing protein information, it will load just one object missing_protein_df
load('cache/missing_protein.rda')
```


# Create regular protein set and missing protein set
create a character vector of missing protein accession only use P2-4 proteins from nextprot
the regular protein is not the just a complementary of missing protein, need to filter P2-5 proteins

```{r}
# missing_proteins is a character vector of MP gene symbols
# similar P2_5_proteins are protein gene symbols under P2-5 category in nextprot
# P2_5_proteins will be used for negative filtering to get regular proteins
missing_proteins_gene_symnol <- missing_protein_df %>% filter(PE != 'Uncertain') %>% select(`gene name(s)`) %>% pull
P2_5_proteins_gene_symbol <- missing_protein_df  %>% select(`gene name(s)`) %>% pull

# missing_proteins is a character vector of MP accession ID
# similar P2_5_proteins are protein accesion ID under P2-5 category in nextprot
missing_proteins_accession <- missing_protein_df %>% filter(PE != 'Uncertain') %>% select(Accession) %>% pull
P2_5_proteins_accession <- missing_protein_df %>% select(Accession) %>% pull
```



# check the number of SpC for each protein for both missing and regular proteins
```{r}
# IP0981_1701 IP0982_1701 IP0993_1701 IP0995_1701 IP0999_1701 IP7100_1701 IP7103_1701 IP7105_1701 sample 1-8
# rename sample to Sample1-8
# colnames(gastric_count) <- c('accession', paste('Sample', 1:8, sep = ' '))

# create long format used for barplot
MP_counts <- gastric_count %>% filter(accession %in% missing_proteins_accession)
MP_counts_long <- MP_counts %>% pivot_longer(names_to = 'IPAS', values_to = 'count', cols = c(-accession)) %>% filter(count > 0)
MP_counts_long$type <- 'Missing'

RP_counts <-  gastric_count %>% filter(!(accession %in% P2_5_proteins_accession))
RP_counts_long <- RP_counts %>% pivot_longer(names_to = 'IPAS', values_to = 'count', cols = c(-accession)) %>% filter(count > 0)
RP_counts_long$type <- 'Regular'

all_counts_long <- rbind(MP_counts_long, RP_counts_long)
```




# Figure S4 The peptide-spectral-matching (PSM) count distribution for regular proteins (Left blue) and missing proteins (Right red)
Frequency of Spectral count for regular proteins
```{r}
# collapse SpC >= 50 into one single category "50+"
RP_barplot_data <-  RP_counts_long %>% mutate(count_discrete=ifelse(count >= 50, '50+', count))
RP_barplot_data$count_discrete <- factor(RP_barplot_data$count_discrete, levels=c(as.character(1:49), '50+'))
RP_barplot_data %>% 
  ggplot(aes(x=count_discrete)) +
  geom_bar(position = 'identity', fill='blue') +
theme_bw() +
  xlab('Spectral counts') +
  ylab('Frequency') +
  ggtitle('Spectral count frequency of regular proteins') +
  scale_x_discrete(breaks=c('1', seq(5, 50, 5), '50+'))
  
  
```



```{r}
MP_barplot_data <-  MP_counts_long %>% mutate(count_discrete=ifelse(count >= 50, '50+', count))
MP_barplot_data$count_discrete <- factor(MP_barplot_data$count_discrete, levels=c(as.character(1:49), '50+'))
MP_barplot_data %>% 
  ggplot(aes(x=count_discrete)) +
  geom_bar(position = 'identity', fill='red') +
theme_bw() +
  xlab('Spectral counts') +
  ylab('Frequency') +
  ggtitle('Spectral count frequency of missing proteins') +
  scale_x_discrete(breaks=c('1', seq(5, 50, 5), '50+'), drop=FALSE)
  
  
  
```




## check the number of SpC for missing proteins
```{r}
MP_counts_long %>% filter(count >= 2) %>% summary()
```


# Match mRNA expression to corresponding protein
The oirginal proteomics data is in accession level and the original transcription data
is in gene level. To match the two data, we need to aggregate protein expression at
the gene level. Typically, one gene can be transcripted and translated into multiple proteins and 
a protein can be translated only from a single gene. So the mapping from accession to gene symbol is unique.
After mapping, there can be rows with the same gene symbol ID, then those row with the 
same gene symbol ID were summed into a single row to represent the expression of proteins translated 
from that one gene.




# create a dictionary used for mapping ensembl_gene_id to gene symbol
When a same ensembl_gene_id can be map to multiple gene symbol (not common)
use only the first one (ordered by alphabeta)
```{r}
dict_accession_to_gene_symbol <- omics_metadata %>% dplyr::select(accession_id, gene_symbol) %>% unique() %>% filter(!is.na(accession_id))
```


# make long-format gastric NSAF data
```{r}
gastric_nsaf_df <- as_tibble(gastric_nsaf, rownames='accession')
gastric_nsaf_df_long <- gastric_nsaf_df %>% pivot_longer(names_to = 'IPAS', values_to = 'NSAF', cols = c(-accession))
gastric_nsaf_df_long <- gastric_nsaf_df_long %>% mutate(type=if_else(accession %in% missing_proteins_accession, 'MP', 
                                                      if_else(!(accession %in% P2_5_proteins_accession), 'RP', 'PE5')))
```

# get the correpsongding gene symbol for each protein
```{r}
gastric_nsaf_df_long$gene_symbol <- mapping(gastric_nsaf_df_long$accession, dict_accession_to_gene_symbol)
rm(dict_accession_to_gene_symbol)
```



# load transcription data

## Transform RNA expression data into gene symbol level
The original RNA data is indexed by ensembl_gene_id (ENSG), to used it with
proteomics data at gene symbol level, we need to mapping ensembl_gene_id to
gene symbol.

```{r}
#
load('cache/count_TPM.rda')
```


## create a dictionary used for mapping ensembl_gene_id to gene symbol
When a same ensembl_gene_id can be map to multiple gene symbol (not common)
use only the first one (ordered by alphabeta)
```{r}
dict_ensembl_to_gen_symbol <- omics_metadata %>% select(ensembl_gene_id, gene_symbol) %>% unique()
dict_ensembl_to_gen_symbol <- dict_ensembl_to_gen_symbol %>% dplyr::group_by(ensembl_gene_id) %>% 
      dplyr::arrange(gene_symbol) %>% 
      dplyr::summarise(map_to=first(gene_symbol))
```

### mapping ensembl_gene_id to gene symbol and pre-process data
1. Remove row with 0 in all cells
2. Only keep the 8 gastric primary cell samples
3. aggregate row with the same index by summing across row axis
```{r}
count_TPM_gene_symbol <- count_TPM_mat
rownames(count_TPM_gene_symbol) <- mapping(rownames(count_TPM_mat), dictionary = dict_ensembl_to_gen_symbol)

count_TPM_gene_symbol <- count_TPM_gene_symbol[rowSums(count_TPM_gene_symbol > 0) > 0, ]

# aggregate row with same gene symbol index
temp <- as_tibble(count_TPM_gene_symbol, rownames='symbol') %>% filter(!is.na(symbol)) %>% group_by(symbol) %>% summarise_all(sum)
count_TPM_gene_symbol <- as.matrix(temp[-1])
rownames(count_TPM_gene_symbol) <- temp$symbol

# change the column name the same as in spectral count matrix
colnames(count_TPM_gene_symbol) <- c('IP7103_1701', 'IP0972_1701', 'IP0995_1701', 'IP0982_1701', 'IP0999_1701', 'IP7100_1701', 'IP0993_1701', 'IP0981_1701', 'IP7105_1701')
count_TPM_gene_symbol <- count_TPM_gene_symbol[, colnames(gastric_nsaf)]

# remove row with NA index
count_TPM_gene_symbol <- count_TPM_gene_symbol[!is.na(rownames(count_TPM_gene_symbol)), ]

head(count_TPM_gene_symbol)
nrow(count_TPM_gene_symbol)
# write.csv(as.data.frame(count_TPM_gene_symbol), file='TPM_gastric_cancer_primary_cell.csv')
```


## make long-format of RNA expression data
```{r}
count_TPM_gene_symbol_df <- as_tibble(count_TPM_gene_symbol,  rownames='gene_symbol')
count_TPM_gene_symbol_df_long <- count_TPM_gene_symbol_df %>% pivot_longer(names_to = 'IPAS', values_to = 'TPM', cols = c(-gene_symbol))
```


## join the RNA expression to protein data
```{r}
dat_combined_long <- left_join(gastric_nsaf_df_long, count_TPM_gene_symbol_df_long, by=c('IPAS', 'gene_symbol'))
dat_combined_long$NSAF <- ifelse(is.na(dat_combined_long$NSAF), 0, dat_combined_long$NSAF)
dat_combined_long$TPM <- ifelse(is.na(dat_combined_long$TPM), 0, dat_combined_long$TPM)
```




## Unique Missing protein accession
```{r}
(MP_unique_accession <- dat_combined_long %>% filter(NSAF > 0, TPM > 0, type=='MP') %>% select(accession) %>% pull() %>% unique())
length(MP_unique_accession)
```




# rename to sample 1- 8
```{r}
# colnames(gastric_nsaf_gene_level) <- paste('Sample', 1:8, sep = ' ')
# colnames(count_TPM_gene_symbol) <- paste('Sample', 1:8, sep = ' ')
```


## combine and match RNA and proteomics data into long format
Each row is a protein identified by gene symbol and sample ID primary key: (accession, IPAS) in term of SQL
TPM is the RNA expression of the gene that coding those proteins
NSAF is the combined protein expressions for proteomics translated from that gene

```{r}
combined_long <- dat_combined_long  %>% filter(!(TPM == 0 & NSAF == 0))
```


## Create a helper function for calculate the number of proteins-at-gene-level has non-zero RNA expression
```{r}
#' @title calculate proportion of protein products supported with RNA products
#' @description It product the proportion of each condition based on different settings
#' @param tbl a tibble object, see details in "dplyr"
#' @param missing_protein whether to only include missing proteins
#' @param group_by_IPAS whether to stratified by IPAS samples
#' @param TPM_cutoff float, the cutoff used for RNA TPM value to consider a RNA product to be existing
#' @param NSAF_cutoff float, the cutoff used for RNA TPM value to consider a protein product to be existing
#' @param cases a vector of integer, used to decide whether cases to included for calculation
summarize_by_case <- function(tbl, protein_type, remove_IPAS=NULL, group_by_IPAS=FALSE, TPM_cutoff = 0, NSAF_cutoff = 0, cases = c(1, 2, 4)) {
  
  tbl <- tbl %>% filter(type==protein_type)
  

  if (!is.null(remove_IPAS)) {
    tbl <- tbl %>% filter(!(IPAS %in% remove_IPAS))
  }
  
  tbl <- tbl %>% rowwise() %>%  
    mutate(case=if_else(TPM > TPM_cutoff & NSAF > NSAF_cutoff, 1,
                        if_else(TPM <= TPM_cutoff & NSAF > NSAF_cutoff, 2, 
                                if_else(TPM > TPM_cutoff & NSAF <= NSAF_cutoff, 3, 4)
                        )
    )
    ) %>% 
    filter(case %in% cases) %>% 
    group_by(IPAS, case) %>% 
    summarise(count=n(), ) %>% 
    ungroup(IPAS, case)
  
  if (group_by_IPAS) {
    tbl <- tbl %>% group_by(IPAS)
  }
  
  tbl <- tbl %>%
    group_by(case, .add=TRUE) %>% 
    summarise(count= sum(count)) %>%
    mutate(countT=sum(count)) %>% 
    mutate(per=paste0(round(100*count/countT,2),'%'))
  
  return(tbl)
}

```


# the proportion of PE2-5 in the missing protein dataset
```{r}
table(missing_protein_df$PE)
```


# check missing proteins with corresponding RNA expression 
```{r}
# check the protein and RNA expression for the detected missing proteins
full_table_detected_MP <- combined_long %>% filter(type=='MP') %>% filter(NSAF > 0) %>% arrange(accession, IPAS)

write_csv(full_table_detected_MP, 'output/detected_177_MP_products_with_RNA.csv')
```


############################ association analysis#####################
# Table 2 The proportions of protein products with mRNA products for missing proteins and regular proteins
## Table 2 missing protein part
```{r}
# the coverage rate of RNA product in protein product for missing protein
(a <- summarize_by_case(combined_long, protein_type = 'MP', group_by_IPAS = TRUE, cases = c(1, 2)))
# write_csv(a, file='output/sample_MP_stats.csv')

## overall without removing the outlier
(b <- summarize_by_case(combined_long, protein_type = 'MP', group_by_IPAS = FALSE, cases = c(1, 2)))

## overall with removing the outlier
(c <- summarize_by_case(combined_long, protein_type = 'MP', group_by_IPAS = FALSE, cases = c(1, 2), remove_IPAS = c('IP0981_1701', 'IP7103_1701')))
```



histogram
```{r}
b$IPAS = 'Combined'
c$IPAS = 'Combined w.o. outlier'
dat_hist <- rbind(a, b, c)


dat_hist$case <- factor(dat_hist$case, levels = c('1', '2'), labels = c('RNA (+)', 'RNA (-)'))
dat_hist$case <- relevel(dat_hist$case, 'RNA (+)')

dat_hist$IPAS <- factor(dat_hist$IPAS, levels = c("IP0981_1701", "IP0982_1701", "IP0993_1701", "IP0995_1701", "IP0999_1701",
                                                  "IP7100_1701", "IP7103_1701", "IP7105_1701", "Combined w.o. outlier",  "Combined"))


ggplot(data=dat_hist, aes(x=IPAS, y=count, group=case, fill=case)) +
  labs(title="Coverage rate for prodein product with RNA product in missing proteins", subtitle = '',
        x = "IPAS", y = 'count') +
  geom_bar( stat = 'identity',  position=position_dodge()) +
  geom_text(aes(label=per), color="black",size=3, vjust=-0.5, position = position_dodge(1)) +
  theme(
        axis.text.x = element_text(angle = 45, hjust = 1))

```




## Table 2 non-missing proteins part
The execution of the following chucks of code could be slow for the size of non-missing proteins
Code is split into several chuck otherwise the computation time required is long.
```{r echo=TRUE}
# the coverage rate of RNA product in protein product for non-missing protein
(a <- summarize_by_case(combined_long, protein_type = 'RP', group_by_IPAS = TRUE, cases = c(1, 2)))
```


```{r echo=TRUE}
## overall without removing the outlier
(b <- summarize_by_case(combined_long, protein_type = 'RP', group_by_IPAS = FALSE, cases = c(1, 2)))
```


```{r echo=TRUE}
## overall with removing the outlier
(c <- summarize_by_case(combined_long, protein_type = 'RP', group_by_IPAS = FALSE, cases = c(1, 2), remove_IPAS = c('IP0981_1701', 'IP7103_1701')))
```


```{r echo=TRUE}
# the coverage rate of protein product in RNA product for non-missing protein
summarize_by_case(combined_long, protein_type = 'RP', group_by_IPAS = TRUE, cases = c(1, 3))
```


```{r echo=TRUE}
## overall without removing the outlier
summarize_by_case(combined_long, protein_type = 'RP', group_by_IPAS = FALSE, cases = c(1, 3))
```


```{r echo=TRUE}
## overall with removing the outlier
summarize_by_case(combined_long, protein_type = 'RP', group_by_IPAS = FALSE, cases = c(1, 3), remove_IPAS = c('IP0981_1701', 'IP7103_1701'))
```

## barplot for non-missing proteins
```{r}
b$IPAS = 'Combined'
c$IPAS = 'Combined2'
dat_hist <- rbind(a, b, c)

dat_hist$case <- factor(dat_hist$case, levels = c('1', '2'), labels = c('RNA (+)', 'RNA (-)'))
dat_hist$case <- relevel(dat_hist$case, 'RNA (+)')

dat_hist$IPAS <- factor(dat_hist$IPAS, levels = c("IP0981_1701", "IP0982_1701", "IP0993_1701", "IP0995_1701", "IP0999_1701",
                                                  "IP7100_1701", "IP7103_1701", "IP7105_1701", "Combined",  "Combined2"))

ggplot(data=dat_hist, aes(x=IPAS, y=count, group=case, fill=case)) +
  labs(title="Coverage rate for prodein product with RNA product in non-missing proteins", subtitle = '',
        x = "IPAS", y = 'count') +
  geom_bar( stat = 'identity',  position=position_dodge()) +
  geom_text(aes(label=per), color="black",size=3, vjust=-0.5, position = position_dodge(1)) +
  theme(
        axis.text.x = element_text(angle = 45, hjust = 1))

```


# Data preparation for Figure 3: The scatter plot of protein-mRNA-product matched pairs. 

## generate data that can be used for plot the scatter plot
Four datasets were created:
1) glmm_dat_missing: paired RNA/protein expressions for missing protein
2) glmm_dat_missing_filtered: pair RNA/protein expressions for missing protein and only keep proteins with both valid (>0) RNA and protein expressions
3) glmm_dat_non_missing: paired RNA/protein expressions for regular protein
4) glmm_dat_non_missing_filtered: air RNA/protein expressions for regular proteins and only keep proteins with both valid (>0) RNA and protein expressions

All TPM/NSAF values were log-transformed in all four datasets
```{r}
glmm_dat_missing <- dat_combined_long %>% filter(type=='MP') %>% mutate(TPM = log(TPM + 1), NSAF = log(NSAF + 1))
glmm_dat_missing_filtered <- glmm_dat_missing %>% filter(TPM > 0 & NSAF > 0)


glmm_dat_non_missing <-dat_combined_long %>% filter(type=='RP') %>% mutate(TPM = log(TPM + 1), NSAF = log(NSAF + 1))
glmm_dat_non_missing_filtered <- glmm_dat_non_missing %>% filter(TPM > 0 & NSAF > 0)
```


## Checking Sample heterogeneity
Check whether RNA expression, protein expression, and protein-RNA-expression relationship were
similar between samples

#  Generalization linear mixed model using RNA expression as random effects in different samples
check whether the correlation relationship is different in different samples
```{r}
library(lme4)
glmm_model <- lmer(NSAF ~ TPM + (0 + TPM | IPAS), data = glmm_dat_missing_filtered)
glm_model <- lm(NSAF ~ TPM, data = glmm_dat_missing_filtered)
summary(glmm_model)
```

No significant evidence for support using RNA expression as random effects
```{r}
anova(glmm_model, glm_model)
```


The scatterplot with group-wise linear regression line validates the result
```{r}
ggplot(data      = glmm_dat_missing_filtered,
       aes(x     = TPM,
           y     = NSAF,
           col   = IPAS,
           group = IPAS))+ #to add the colours for different classes
  geom_point(size     = 1.2,
             alpha    = .8,
             position = "jitter")+ #to add some random noise for plotting purposes
  theme_minimal() +
  geom_smooth(method = lm,
              se     = FALSE,
              size   = .5,
              alpha  = .8)+ # to add regression line
  labs(title    = "Normalized spectral abundance factor (NASF) vs. Transcripts per million (TPM)",
       subtitle = "missing proteins on gene level in gastric cancer primary cell samples",
       xlab = 'Log2(TPM) + 1',
       ylab = 'Log2(NSAF) + 1',
  )
```


```{r}
model.lm <- lm(NSAF ~ TPM, data = glmm_dat_missing_filtered)
summary(model.lm)
```



## correlation plot for pairwise correlation of missing protein expression
# results of the two correlation plot can be found in the end of the this markdown file
```{r eval=FALSE, include=FALSE}
library(corrplot)
corr <- cor(heatmap_dat, method = "spearman")
corrplot.mixed(corr, lower = 'number', upper='ellipse', tl.cex=0.5,
               na.label.col='black',
               mar=c(0,0,2,0))
```

## correlation plot for pairwise correlation of missing protein gene expression
```{r eval=FALSE, include=FALSE}
corr <- cor(heatmap_dat_TPM, method = "spearman")
corrplot.mixed(corr, lower = 'number', upper='ellipse', tl.cex=0.6, na.label.col='black')
```



# Figure 3 with spearman correlation for missing proteins
The scatter plot of protein-mRNA-product matched pairs. (a): All protein-mRNA
pairs for missing proteins.  (b): protein-mRNA pairs with TPM > 0 and NSAF > 0. 
(c): All protein-mRNA pairs for regular proteins. (d): All protein-mRNA pairs
for regular proteins with TPM > 0 and NSAF > 0. Blue line: Simple linear 
regression. Spearman correlations and corresponding p-value were calculated for (b) and (d). 
```{r}
par(mfrow=c(2, 2))
ggpubr::ggscatter(glmm_dat_missing, x = "TPM", y = "NSAF",
   color = 'IPAS', size = 1.5, # Points color, shape and size
   title='Plot of NSAF vs. TPM for missing proteins',
   xlab = 'Log2 (TPM + 1)',
   ylab = 'Log2 (NSAF + 1)',
   ) + 
  scale_color_discrete(labels=paste('sample', 1:8)) +
  scale_y_continuous(name='Log2 (NSAF + 1)', breaks=c(0, 2, 4, 6, 8, 10), limits = c(0, 10)) +
  # geom_vline(xintercept=0.1, linetype=2, color='red', size=0.5) +
  geom_hline(yintercept=2, linetype=2, color='red', size=0.5) +
  theme(legend.text = element_text(size=10)) + 
  labs(color='Sample')


# plot with spearman correlation for missing proteins
plt <- ggpubr::ggscatter(glmm_dat_missing_filtered, x = "TPM", y = "NSAF",
   color = 'IPAS', size = 1.5, # Points color, shape and size
   add = "reg.line",  # Add regressin line
   add.params = list(color = "blue", fill = "lightgray"), # Customize reg. line
   conf.int = TRUE, # Add confidence interval
   cor.coef = TRUE, # Add correlation coefficient. see ?stat_cor
   cor.coeff.args = list(method = "spearman", cor.coef.name = "rho", label.x.npc=0.8, label.sep = "\n",
                         label.y.npc = 0.2),
   title='Plot of NSAF vs. TPM for missing proteins with NSAF > 0 and TPM > 0',
   # subtitle='missing proteins on gene level in gastric cancer primary cell sample',
   xlab = 'Log2 (TPM + 1)',
   ylab = 'Log2 (NSAF + 1)',
   ) +
   scale_color_discrete(labels=paste('sample', 1:8)) +
   scale_y_continuous(name='Log2 (NSAF + 1)', breaks=c(0, 2, 4, 6, 8, 10), limits = c(0, 8)) +
   theme(legend.text = element_text(size=10)) +
   labs(color='Sample')

dens1 <- ggplot(glmm_dat_missing_filtered, aes(x = TPM, fill = IPAS)) + 
  geom_density(alpha = 0.4) + 
  theme_void() + 
  theme(legend.position = "none")

dens2 <- ggplot(glmm_dat_missing_filtered, aes(x = NSAF, fill = IPAS)) + 
  geom_density(alpha = 0.4) + 
  theme_void() + 
  theme(legend.position = "none") + 
  coord_flip()

dens1 + patchwork::plot_spacer() + plt + dens2 +  patchwork::plot_layout(ncol = 2, nrow = 2, widths = c(4, 1), heights = c(1, 4))


# plot with spearman correlation for non-missing proteins
ggpubr::ggscatter(glmm_dat_non_missing, x = "TPM", y = "NSAF",
   color = 'IPAS', size = 1.5, # Points color, shape and size
   title='Plot of NSAF vs. TPM for regular proteins',
   xlab = 'Log2 (TPM + 1)',
   ylab = 'Log2 (NSAF + 1)') +
  scale_color_discrete(labels=paste('sample', 1:8)) +
   scale_x_continuous(name='Log2 (TPM + 1)', breaks=seq(0, 16, 2), limits = c(0, 16)) +
   scale_y_continuous(name='Log2 (NSAF + 1)', breaks=seq(0, 14, 2), limits = c(0, 14)) +
   theme(legend.text = element_text(size=10)) + 
  # geom_vline(xintercept=0.1, linetype=2, color='red', size=0.5) +
  geom_hline(yintercept=2, linetype=2, color='red', size=0.5) +
  labs(color='Sample')

plt <- ggpubr::ggscatter(glmm_dat_non_missing_filtered, x = "TPM", y = "NSAF",
   color = 'IPAS', size = 1.5, # Points color, shape and size
   add = "reg.line",  # Add regressin line
   add.params = list(color = "blue", fill = "lightgray"), # Customize reg. line
   conf.int = TRUE, # Add confidence interval
   cor.coef = TRUE, # Add correlation coefficient. see ?stat_cor
   cor.coeff.args = list(method = "spearman", cor.coef.name = "rho",  label.x.npc=0.85, label.sep = "\n",
                         label.y.npc = 0.2),
   title='Plot of NSAF vs. TPM for regular proteins with NSAF > 0 and TPM > 0',
   # subtitle='missing proteins on gene level in gastric cancer primary cell sample',
   xlab = 'Log2 (TPM + 1)',
   ylab = 'Log2 (NSAF + 1)') +
  scale_color_discrete(labels=paste('sample', 1:8)) +
   scale_x_continuous(name='Log2 (TPM + 1)', breaks=seq(0, 16, 2), limits = c(0, 16)) +
   scale_y_continuous(name='Log2 (NSAF + 1)', breaks=seq(0, 14, 2), limits = c(0, 14)) +
    theme(legend.text = element_text(size=10)) +
  labs(color='Sample')

dens1 <- ggplot(glmm_dat_non_missing_filtered, aes(x = TPM, fill = IPAS)) + 
  geom_density(alpha = 0.4) + 
  theme_void() + 
  theme(legend.position = "none")

dens2 <- ggplot(glmm_dat_non_missing_filtered, aes(x = NSAF, fill = IPAS)) + 
  geom_density(alpha = 0.4) + 
  theme_void() + 
  theme(legend.position = "none") + 
  coord_flip()

dens1 + patchwork::plot_spacer() + plt + dens2 +  patchwork::plot_layout(ncol = 2, nrow = 2, widths = c(4, 1), heights = c(1, 4))
```


# Figure 2: Heatmap for missing protein

Extract the 85 identified unique missing proteins
Criterion: has both valid protein expression and RNA expression i.e.  NASF > 0
& TPM > 0
```{r}
# the number should be 85 if using two-SpC-rule
detected_MP_proteins <- combined_long  %>% filter(NSAF > 0, TPM > 0, type=='MP') %>% arrange(accession) %>% group_by(accession) %>% summarise(n=n())  %>%  select(accession) %>% pull() %>% unique()
# the details of the 85 identified missing proteins
detected_MP_proteins

detected_MP_proteins_gene_symol <- combined_long  %>% filter(NSAF > 0, TPM > 0, type=='MP') %>% select(gene_symbol) %>% pull() %>% unique()
```
# the PE (protein evidence) distribution of the identified missing proteins
```{r}
missing_protein_df %>% filter(`Accession` %in% detected_MP_proteins) %>% select(PE) %>% pull() %>% table()
```


## Figure 2A) 
Heatmap representations for quantified missing proteins in the eight 
gastric cancer primary cell samples. Color represents the level of protein 
expression, and values were log (NSAF +1) transformed
```{r, fig.height=8, fig.width=8}
# double check clustering
library(pheatmap)
heatmap_dat <- log(gastric_nsaf[detected_MP_proteins, ] + 1)
colnames(heatmap_dat) =paste('sample', 1:8)
pheatmap(heatmap_dat, 
         show_rownames = T,
         angle_col=45,
         cluster_cols = T,
         cluster_rows = T)
```

## Figure 2 additional
Heatmap representations for quantified missing proteins in the eight 
gastric cancer primary cell samples. Color represents the level of mRNA expression,
and values were log (TPM +1) transformed
```{r, fig.height=12, fig.width=8}
heatmap_dat_TPM <- log(count_TPM_gene_symbol[detected_MP_proteins_gene_symol, ] + 1)
colnames(heatmap_dat_TPM) =paste('sample', 1:8)
pheatmap(heatmap_dat_TPM, 
         show_rownames = T,
         angle_col=45,
         cluster_cols = T,
         cluster_rows = T)
```




## correlation plot for pairwise correlation of missing protein expression
```{r}
library(corrplot)
corr <- cor(heatmap_dat, method = "spearman")
corrplot.mixed(corr, lower = 'number', upper='ellipse', tl.cex=0.5,
               na.label.col='black',
               mar=c(0,0,2,0))
```

## correlation plot for pairwise correlation of missing protein gene expression
```{r}
corr <- cor(heatmap_dat_TPM, method = "spearman")
corrplot.mixed(corr, lower = 'number', upper='ellipse', tl.cex=0.6, na.label.col='black')
```


