---
title: "RNA data processing and exploration"
output:
  pdf_document: default
  html_document: default
date: "2022-09-27"
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
knitr::opts_knit$set(root.dir = '/Volumes/hanash/Hanash_Lab_GPFS/Chunhui/missing protien project manuscript/missing-protein-project-data')
# run following line if you need to run code in console
# knitr::opts_knit$set(root.dir only change working directory of code chucks not code directly running in console
# setwd('/Users/cgu3/Library/CloudStorage/OneDrive-InsideMDAnderson/proteomics/missing protien project manuscript/missing-protein-project-datat')
# getwd()
```



```{r}
library(tidyverse)
if (!require("protools", quietly = TRUE))
   devtools::install_github("https://github.com/FDUguchunhui/protools")
library(protools)
```


# import the RNA data
```{r}
count <- read_csv('RNA-data/merged_IPAS_RNAseq_counts_gastric.csv')
count <- count[-c(1:5), ]
colnames(count) <- c('Ensemble_ID', colnames(count)[-1])
count$Ensemble_ID <-str_extract(count$Ensemble_ID, pattern = '.+(?=\\.)')
head(count)
```


# data wrangling
combine duplicate ensemble_id
63,920 -> 63,875
```{r}
count <- count %>% group_by(Ensemble_ID) %>% summarise_all(sum)
```


63,875 -> 43,939
remove rows that all cells are 0
```{r}
count <- count[-which(rowSums(count[-1]) == 0),]
```




# import missing protein information
```{r}
## read-in newest missing protein list
missing_protein_df <- readxl::read_xlsx('support-data/PE2-5.xlsx')
missing_protein_df <- missing_protein_df[-c(1:12), ]
head(missing_protein_df)
```



# Section 0: Exploration
## overall distribution
```{r}
count_df <- as.data.frame(count[-1])
pseudo_counts <- log(count_df + 1)
ggplot(data = data.frame(x=c(as.matrix(pseudo_counts))), aes(x=x)) + 
  geom_histogram(aes(y = ..density..),
                 colour = 1, fill = "white") +
  geom_density(lwd = 1, colour = 4,
               fill = 4, alpha = 0.25)
```


## tag-wise distribution
```{r}
sample_filter <- sample(1:nrow(count_df), 100)
density_plot_dat <- t(count_df[sample_filter,]) %>% as_tibble() %>% pivot_longer(cols = everything(), names_to = 'ensemble_id', values_to = 'count') 
plot <- ggplot(data = density_plot_dat, aes(x=count, color=ensemble_id)) + 
  geom_density(alpha = 0.2) +
  scale_x_continuous(breaks=0:10, limits=c(0, 10)) +
  theme(legend.position="none")
suppressWarnings(print(plot))
```

## sample-wise distribution
```{r}
boxplot(log(count_df+1))
```




# Section 1 data processing
## do the conversion using biomaRt
the result is better than using org.Hs.eg.db
no missing gene symbol

! there are some Ensembl gene id cannot be found by biomart
https://support.bioconductor.org/p/111608/
```{r}
# Uncomment the following code if you want to get the dictionary for mapping ensemble ID to gene symbol from the source
# mart <- biomaRt::useDataset("hsapiens_gene_ensembl", useMart("ensembl"))
# dict_ensembl_to_symbol <- biomaRt::getBM(filters= "ensembl_gene_id", attributes= c("ensembl_gene_id","hgnc_symbol", "start_position","end_position"), values=count$Ensemble_ID, mart= mart)
# dict_ensembl_to_symbol$gene_length=dict_ensembl_to_symbol$end_position - dict_ensembl_to_symbol$start_position
# dict_ensembl_to_symbol[dict_ensembl_to_symbol == ''] <- NA
# save(dict_ensembl_to_symbol, file='cache/dict_ensemble_to_symbol.rda')
load('cache/dict_ensemble_to_symbol.rda')
head(dict_ensembl_to_symbol)
write_csv(dict_ensembl_to_symbol, 'support-data/dict_ensembl_to_symbol.csv')

# code below can be used to check those unmapped ensembl gene id
# count$Ensemble_ID[!count$Ensemble_ID %in% dict_ensembl_to_symbol$ensembl_gene_id]
```


## filtering unmapped ensemble gene id
```{r}
count <- count %>% filter(Ensemble_ID %in% dict_ensembl_to_symbol$ensembl_gene_id)
```



## recreate data.frame and matrix
after several necessary pre-processing create data.frame and matrix version of the data
```{r}
count_df <- as.data.frame(count[-1])
rownames(count_df) <- count$Ensemble_ID
count_mat <- as.matrix(count_df)
```



## normalization: calculate  TPM (transcripts per million)
```{r}
gene_length_df <- dict_ensembl_to_symbol %>% dplyr::select(ensembl_gene_id, gene_length) %>% unique()
gene_length_df <- data.frame(length=gene_length_df$gene_length, row.names=gene_length_df$ensembl_gene_id)
count_TPM_mat <- TPM(count_mat, gene_length_df)
head(count_TPM_mat)
# save(count_TPM_mat, file='cache/count_TPM.rda')
```


# Transform RNA expression data into gene symbol level
The original RNA data is indexed by ensembl_gene_id (ENSG), to used it with
proteomics data at gene symbol level, we need to mapping ensembl_gene_id to
gene symbol.

## create a dictionary used for mapping ensembl_gene_id to gene symbol
When a same ensembl_gene_id can be map to multiple gene symbol (not common)
use only the first one (ordered by alphabeta)
```{r}
# import information for mapping between accession, gene symbol, and ensembl ID
omics_metadata <- read_csv('support-data/omics_metadata.csv', show_col_types = FALSE)

dict_ensembl_to_gen_symbol <- omics_metadata %>% select(ensembl_gene_id, gene_symbol) %>% unique()
dict_ensembl_to_gen_symbol <- dict_ensembl_to_gen_symbol %>% dplyr::group_by(ensembl_gene_id) %>% 
      dplyr::arrange(gene_symbol) %>% 
      dplyr::summarise(map_to=first(gene_symbol))

rm(omics_metadata)
```

## mapping ensembl_gene_id to gene symbol and pre-process data
1. Remove row with 0 in all cells
2. Only keep the 8 gastric primary cell samples
3. aggregate row with the same index by summing across row axis
```{r}
count_TPM_gene_symbol <- count_TPM_mat
# for rowname, mapping ensembl ID to gene symbol
rownames(count_TPM_gene_symbol) <- mapping(rownames(count_TPM_mat), 
                                           dictionary = dict_ensembl_to_gen_symbol)
# remove all 0 rows
count_TPM_gene_symbol <- count_TPM_gene_symbol[rowSums(count_TPM_gene_symbol > 0) > 0, ]

# aggregate row with same gene symbol index
temp <- as_tibble(count_TPM_gene_symbol, rownames='symbol') %>% filter(!is.na(symbol)) %>% group_by(symbol) %>% summarise_all(sum)
count_TPM_gene_symbol <- as.matrix(temp[-1])
rownames(count_TPM_gene_symbol) <- temp$symbol

# change the column name the same as in spectral count matrix
colnames(count_TPM_gene_symbol) <- c('IP7103_1701', 'IP0972_1701', 'IP0995_1701', 'IP0982_1701', 'IP0999_1701', 'IP7100_1701', 'IP0993_1701', 'IP0981_1701', 'IP7105_1701')
count_TPM_gene_symbol <- count_TPM_gene_symbol[, c("IP0981_1701", "IP0982_1701",
                                                   "IP0993_1701", "IP0995_1701",
                                                   "IP0999_1701", "IP7100_1701",
                                                   "IP7103_1701", "IP7105_1701")]

# remove row with NA index
count_TPM_gene_symbol <- count_TPM_gene_symbol[!is.na(rownames(count_TPM_gene_symbol)), ]

head(count_TPM_gene_symbol)
nrow(count_TPM_gene_symbol)
# write.csv(as.data.frame(count_TPM_gene_symbol), file='TPM_gastric_cancer_primary_cell.csv')
```


# export the processed RNA data
```{r}
write.csv(as.data.frame(count_TPM_gene_symbol), 'RNA-data/processed_RNA_data.csv')
```


## recheck sample-level expression distribution after TPM normalization
```{r}
boxplot(as.data.frame(log(count_TPM_gene_symbol+1)))
```


